openpower/patches/vesnin-patches/hostboot-p8/hostboot-1004-Support-for-extended-FRU-data-for-memory-modules.patch

```diff
@@ -0,0 +1,450 @@
++            js = 0;
++    }
++
++    for (size_t i = 0; i < js; ++i) {
++        if (jm[i].id == jedecId) {
++            decoded = jm[i].decoded;
++            break;
++        }
++    }
++
++    return decoded;
++}
++
++
++/**
++ * @brief Put string data into IPMI message buffer.
++ *
++ * @param[out] msgData, IPMI message buffer
++ * @param[in] buff, Source buffer containing text string
++ * @param[in] len, Length of buffer in bytes without null-termination symbol
++ */
++static void putIpmiString(std::vector<uint8_t>& msgData,
++                          const char* buff,
++                          size_t len)
++{
++    // Trim string to maximum lenght allowed
++    if (len > IPMIFRUINV::MAX_ASCII_FIELD_SIZE)
++        len = IPMIFRUINV::MAX_ASCII_FIELD_SIZE;
++
++    msgData.push_back(len + IPMIFRUINV::TYPELENGTH_BYTE_ASCII);
++    msgData.insert(msgData.end(), buff, buff + len);
++}
++
++
++/**
++ * @brief Read value from SPD field and decode it.
++ *
++ * @param[in] target, Descriptor of target module to read
++ * @param[in] spdId, SPD field identifier
++ * @param[out] value, SPD field value
++ * @param[out] decoded, Pointer to save decoded value, optional
++ *
++ * @return bool, true if operation was completed successfully
++ */
++static bool readSpdField(TARGETING::TargetHandle_t target,
++                         uint16_t spdId,
++                         uint8_t& value,
++                         const char** decoded = NULL)
++{
++    size_t fieldSz = sizeof(value);
++    if (deviceRead(target, &value, fieldSz, DEVICE_SPD_ADDRESS(spdId) != NULL))
++        return false; // Read error
++    if (decoded)
++        *decoded = jedecDecode(spdId, value);
++    return true;
++}
++
++
++void memModuleManufacturer(TARGETING::TargetHandle_t target,
++                           std::vector<uint8_t>& msgData)
++{
++    const uint16_t spdId = SPD::MODULE_MANUFACTURER_ID;
++
++    // Read manufacturer Id, ignore errors as FRU is not a critical operation
++    uint16_t manufacturerId = 0;
++    size_t sz = sizeof(manufacturerId);
++    deviceRead(target, &manufacturerId, sz, DEVICE_SPD_ADDRESS(spdId));
++    manufacturerId = le16toh(manufacturerId); // SPD values are stored in LE format
++
++    const char* name = jedecDecode(spdId, manufacturerId);
++    if (name)
++        putIpmiString(msgData, name, strlen(name));
++    else {
++        char buf[32];
++        const size_t len = snprintf(buf, sizeof(buf),
++                                    "Unknown (0x%04x)", manufacturerId);
++        putIpmiString(msgData, buf, len);
++    }
++}
++
++
++void memModuleDetails(TARGETING::TargetHandle_t target,
++                      std::vector<uint8_t>& msgData)
++{
++    // Read detailed memory module info,
++    // ignore errors as FRU is not a critical operation
++    uint8_t basicType = 0xff;
++    const char* basicTypeName = NULL;
++    readSpdField(target, SPD::BASIC_MEMORY_TYPE, basicType, &basicTypeName);
++    uint8_t moduleType = 0xff;
```

Comment:
Please, for readability, add blank lines between each `readSpdField` and next variable declaration.